merkle-tree/concurrent/src/lib.rs

```diff
@@ -0,0 +1,317 @@
+    ) -> ([u8; 32], [[u8; 32]; MAX_HEIGHT]) {
+        let mut updated_leaf = *old_leaf;
+        let mut updated_proof = proof.to_owned();
+        let mut k = changelog_index;
+
+        while k != self.current_changelog_index as usize {
+            let changelog_entry = self.changelog[k];
+            let changelog_entry_index = changelog_entry.index as usize;
+            if leaf_index != changelog_entry_index {
+                let intersection_index = self.intersection_index(leaf_index, changelog_entry_index);
+                updated_proof[intersection_index] = changelog_entry.path[intersection_index];
+            } else {
+                updated_leaf = changelog_entry.path[0];
+            }
+
+            k = (k + 1) % MAX_ROOTS;
+        }
+
+        (updated_leaf, updated_proof)
+    }
+
+    /// Updates the leaf under `leaf_index` with the `new_leaf` value.
+    ///
+    /// # Validation
+    ///
+    /// This method doesn't validate the proof. Caller is responsible for
+    /// doing that before.
+    fn update_leaf_in_tree(
+        &mut self,
+        new_leaf: &[u8; 32],
+        leaf_index: usize,
+        proof: &[[u8; 32]],
+    ) -> Result<(), HasherError> {
+        let mut node = *new_leaf;
+        let mut changelog_path = [[0u8; 32]; MAX_HEIGHT];
+
+        for (j, sibling) in proof.iter().enumerate() {
+            changelog_path[j] = node;
+            node = compute_parent_node::<H>(&node, sibling, leaf_index, j)?;
+        }
+
+        let changelog = ChangelogEntry::new(node, changelog_path, leaf_index);
+        self.inc_current_changelog_index();
+        self.changelog[self.current_changelog_index as usize] = changelog;
+        self.inc_current_root_index();
+        self.roots[self.current_root_index as usize] = node;
+
+        Ok(())
+    }
+
+    /// Replaces the `old_leaf` under the `leaf_index` with a `new_leaf`, using
+    /// the given `proof` and `changelog_index` (pointing to the changelog entry
+    /// which was the newest at the time of preparing the proof).
+    pub fn update(
+        &mut self,
+        changelog_index: usize,
+        old_leaf: &[u8; 32],
+        new_leaf: &[u8; 32],
+        leaf_index: usize,
+        proof: &[[u8; 32]],
+    ) -> Result<(), HasherError> {
+        // Provided proof might have lower number of elements than the height.
+        // In such case, fill the proof with zero bytes.
+        let proof: [[u8; 32]; MAX_HEIGHT] = if proof.len() < MAX_HEIGHT {
+            full_proof::<H, MAX_HEIGHT>(proof)
+        } else {
+            proof.try_into().map_err(|_| HasherError::ProofTooLarge)?
+        };
+
+        let (updated_leaf, updated_proof) =
+            self.update_proof_and_leaf(changelog_index, old_leaf, leaf_index, &proof);
+
+        if is_valid_proof::<H, MAX_HEIGHT>(
+            &self.roots[self.current_root_index as usize],
+            old_leaf,
+            leaf_index,
+            &proof,
+        )? && updated_leaf == *old_leaf
```

Comment:
why do we update the leaf and then assert it is equal to the old leaf. So we don't actually want to update the leaf?
Then we could also just throw an error once we notice that we would need to update the leaf right?

Also if the updated_leaf is not equal to old leaf we don't throw an error and just return ok.
We should throw an error right?

Reply:
I'm returning an error now.

The `updated_leaf` was somehow confusing (for my defence - took from the whitepaper :stuck_out_tongue: ) - what it meant was "the new value of the leaf which you are attempting to replace, which comes from a new changelog".

But it's unnecessary. I removed it completely. I'm just returning an error whenever we find a changelog affecting the replaced leaf directly.